pkg/util/translate/translate.go

```diff
@@ -252,6 +254,44 @@ func ResetObjectMetadata(obj metav1.Object) {
 	obj.SetManagedFields(nil)
 }
 
+type ApplyMetadataOptions struct {
+	SetHostNameAndNamespaceAnnotations bool
+	ExcludeAnnotations                 []string
+}
+
+// SyncHostMetadataToVirtual copies metadata from the host resource to the virtual resource.
+//
+// In addition to copying the annotations from the host resource, this function also sets multiple vCluster annotations
+// on the virtual resource:
+//   - vcluster.loft.sh/managed-annotations: a multi-line string that contains keys of annotations that are copied from
+//     the host resource;
+//   - vcluster.loft.sh/managed-labels: a multi-line string that contains keys of labels that are copied from the host
+//     resource;
+//   - vcluster.loft.sh/object-host-name: host resource name, so you can easily identify from which host resource the
+//     virtual resource was created;
+//   - vcluster.loft.sh/object-host-namespace: host resource namespace (if the resource is namespaced).
+//
+// The following annotations are not copied from the host resource:
+//   - vcluster.loft.sh/managed-annotations
+//   - vcluster.loft.sh/managed-labels
+//
+// It also sets the following labels on the virtual resource:
+//   - vcluster.loft.sh/managed-by: vCluster name, so you can know that the resource is managed by vCluster.
+//
+// When this function exists, virtual resource has annotations and labels maps created, even if they are empty.
+func SyncHostMetadataToVirtual(pObj client.Object, vObj client.Object, options ApplyMetadataOptions) {
+	labels, annotations := ApplyMetadata(pObj.GetAnnotations(), vObj.GetAnnotations(), pObj.GetLabels(), vObj.GetLabels(), options.ExcludeAnnotations...)
+	labels[MarkerLabel] = VClusterName
+	if options.SetHostNameAndNamespaceAnnotations {
+		annotations[HostNameAnnotation] = pObj.GetName()
+		if pObj.GetNamespace() != "" {
+			annotations[HostNamespaceAnnotation] = pObj.GetNamespace()
+		}
+	}
+	vObj.SetAnnotations(annotations)
+	vObj.SetLabels(labels)
+}
```

Comment:
Unfortunately we cannot use this approach as we cannot rely on labels or annotations within the vCluster as the end user can freely change this